### PR TITLE
Sms prefix billing

### PR DESCRIPTION
**Pull request by nickpell**
implements http://manage.dimagi.com/default.asp?154612#877396

@gcapalbo @biyeun

### CODE REVIEW

File: corehq/apps/smsbillables/models.py [hunk 1]
```diff
@@ -27,23 +28,46 @@ class SmsGatewayFeeCriteria(models.Model):
     backend_instance = models.CharField(max_length=255, db_index=True, null=True)
     direction = models.CharField(max_length=10, db_index=True, choices=DIRECTION_CHOICES)
     country_code = models.IntegerField(max_length=5, null=True, blank=True, db_index=True)
+    prefix = models.CharField(max_length=10, blank=True, default="", db_index=True)
 
     @classmethod
-    def get_most_specific(cls, backend_api_id, direction, backend_instance=None, country_code=None):
+    def get_most_specific(cls, backend_api_id, direction, backend_instance=None, country_code=None, national_number=None):
```
Comment (dimagimon) [inline corehq/apps/smsbillables/models.py]: E501 line too long (122 > 115 characters)

File: corehq/apps/smsbillables/models.py [hunk 2]
```diff
@@ -94,10 +137,14 @@ def create_new(cls, backend_api_id, direction, amount,
 
 
     @classmethod
-    def get_by_criteria(cls, backend_api_id, direction, backend_instance=None, country_code=None):
-        criteria = SmsGatewayFeeCriteria.get_most_specific(backend_api_id, direction,
-                                                           backend_instance=backend_instance,
-                                                           country_code=country_code)
+    def get_by_criteria(cls, backend_api_id, direction, backend_instance=None, country_code=None, national_number=None):
```
Comment (dimagimon) [inline corehq/apps/smsbillables/models.py]: E501 line too long (120 > 115 characters)

**Comment by dimagimon** [comment]
:+1: No lint errors found.

**Comment by nickpell** [comment]
@gcapalbo this is good to go when you're ready

File: corehq/apps/smsbillables/management/commands/bootstrap_grapevine_gateway.py [hunk 3]
```diff
@@ -5,37 +5,50 @@
...
+    SmsGatewayFee.create_new(
+        GrapevineBackend.get_api_id(), INCOMING, Decimal('0.10'),
+        currency=currency,
+        fee_class=sms_gateway_fee_class,
+        criteria_class=sms_gateway_fee_criteria_class,
+    )
+    logger.info("Updated Global Grapevine gateway fees.")
+
+    # messages relevant to our Grapevine Backends
+    for backend in relevant_backends:
```
Comment (gcapalbo) [inline corehq/apps/smsbillables/management/commands/bootstrap_grapevine_gateway.py]: How come we iterate over this list? There's only one grapevine backend we have hooked up, and we don't want to tie rates to the instance for grapevine, only the api id

File: corehq/apps/smsbillables/management/commands/bootstrap_tropo_gateway.py [hunk 4]
```diff
@@ -1,33 +1,59 @@
...
+        if data[1] == 'Fixed Line' and data[4] != '\n':
+            SmsGatewayFee.create_new(
+                TropoBackend.get_api_id(),
+                OUTGOING,
+                float(data[4].rstrip()),
+                country_code=int(data[2]),
+                currency=currency,
+                fee_class=sms_gateway_fee_class,
+                criteria_class=sms_gateway_fee_criteria_class,
+            )
```
Comment (gcapalbo) [inline corehq/apps/smsbillables/management/commands/bootstrap_tropo_gateway.py]: Looks like this is missing inbound rates for Tropo per country?

File: corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py [hunk 5]
```diff
@@ -7,101 +7,124 @@
...
+                if not(iso in mach_data):
+                    mach_data[iso] = {}
+                mach_data[iso][network] = (country_code, subscribers)
+                row += 1
+        except IndexError:
+            pass
+        return mach_data
+
+    twilio_data = get_twilio_data()
+    mach_data = get_mach_data()
```
Comment (gcapalbo) [inline corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py]: I don't understand the use of the mach_data here - what's it used for?

File: corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py [hunk 6]
```diff
@@ -7,101 +7,124 @@
...
+                weighted_price = weighted_price / total_subscriptions
+                SmsGatewayFee.create_new(
+                    TwilioBackend.get_api_id(),
+                    OUTGOING,
+                    weighted_price,
+                    country_code=country_code,
+                    currency=currency_class.objects.get(code="USD"),
+                    fee_class=sms_gateway_fee_class,
+                    criteria_class=sms_gateway_fee_criteria_class,
+                )
```
Comment (gcapalbo) [inline corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py]: Looks like Twilio is missing inbound rates per country?

File: corehq/apps/smsbillables/models.py [hunk 7]
```diff
@@ -27,23 +28,47 @@ class SmsGatewayFeeCriteria(models.Model):
...
+        - country_code and prefix (optional)
         """
         all_possible_criteria = cls.objects.filter(backend_api_id=backend_api_id, direction=direction)
 
         if all_possible_criteria.count() == 0:
             return None
 
+        national_number = str(national_number) if national_number is not None else ""
+        if national_number:
+            national_number = '0' * (10 - len(national_number)) + national_number
```
Comment (gcapalbo) [inline corehq/apps/smsbillables/models.py]: Why are we padding the national_number with leading zeroes here?

File: corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+    )
+
+    backend_id = '7ddf3301c093b793c6020ebf755adb6f'
+    try:
+        backend = HttpBackend.get(backend_id)
+    except ResourceNotFound:
+        # for non-production environments
+        backend = HttpBackend()
+        backend._id = backend_id
+        backend.save()
```
Comment (gcapalbo) [inline corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py]: We shouldn't create a backend like this - if the backend is not found then I think we should just do nothing and have a warning message that gets displayed

**Comment by gcapalbo** [comment]
@nickpell - Made some comments - also, does this need any testing or qa before merging?

File: corehq/apps/smsbillables/models.py (hunk 7 above)
Comment (nickpell) [inline corehq/apps/smsbillables/models.py]: the phone number parsing library returns a numerical value instead of a string, so leading zeroes are dropped

File: corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py (hunk 6 above)
Comment (nickpell) [inline corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py]: that's correct - I'd recommend adding them in a new migration, since we don't want to re-run migrations on every running instance of hq

File: corehq/apps/smsbillables/management/commands/bootstrap_tropo_gateway.py (hunk 4 above)
Comment (nickpell) [inline corehq/apps/smsbillables/management/commands/bootstrap_tropo_gateway.py]: see comment regarding Twilio

File: corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py (hunk 5 above)
Comment (nickpell) [inline corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py]: Taking a quick glance, I don't remember exactly how I implemented this before ... but there's no changes to these management commands other than ensuring when the ORM changes, so I'm not too concerned.  If you want more details I can take a closer look.

File: corehq/apps/smsbillables/management/commands/bootstrap_grapevine_gateway.py (hunk 3 above)
Comment (nickpell) [inline corehq/apps/smsbillables/management/commands/bootstrap_grapevine_gateway.py]: same response as in the twilio bootstrap - I'm not sure about this legacy code, but I can look further if you want

File: corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py (hunk 8 above)
Comment (nickpell) [inline corehq/apps/smsbillables/management/commands/bootstrap_moz_gateway.py]: fixed

**Comment by nickpell** [comment]
@gcapalbo Replied to comments.  I'm ok QAing or testing in production

File: corehq/apps/smsbillables/models.py (hunk 7 above)
Comment (gcapalbo) [inline corehq/apps/smsbillables/models.py]: I see - but so then isn't this code assuming that the length of the national number is always 10 (I don't think that's a safe assumption to make)?

File: corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py (hunk 5 above)
Comment (gcapalbo) [inline corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py]: ok - I just want to make sure we're not mixing rate data - can we make sure we're not pulling mach rate data into the twilio rates?

File: corehq/apps/smsbillables/management/commands/bootstrap_grapevine_gateway.py (hunk 3 above)
Comment (gcapalbo) [inline corehq/apps/smsbillables/management/commands/bootstrap_grapevine_gateway.py]: We should fix this so that we're only tying grapevine rates to the api_id and not the instance, but I think it's also fine to do that in another pr. Also can we doublecheck these rates? I'd be surprised if outbound costed more than inbound for this one

File: corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py (hunk 5 above)
Comment (nickpell) [inline corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py]: yes - what's going on here is it's using Mach's statistics of the breakdown of the number of subscribers to carriers to guess the percentage breakdown of the carriers that Twilio messages go to, which in turn is used to estimate the average gateway fee for a twilio message to a particular country.  A lot of guesswork going on, but it's the best we have at the moment.

File: corehq/apps/smsbillables/models.py (hunk 7 above)
Comment (nickpell) [inline corehq/apps/smsbillables/models.py]: I am assuming that - do you think that will cause problems?

File: corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py (hunk 5 above)
Comment (gcapalbo) [inline corehq/apps/smsbillables/management/commands/bootstrap_twilio_gateway.py]: I see, ok that makes sense

File: corehq/apps/smsbillables/models.py (hunk 7 above)
Comment (gcapalbo) [inline corehq/apps/smsbillables/models.py]: Yeah, we can't make that assumption.  Can't we just take the original phone number, subtract the country code from it, and use that when matching the prefix?

File: corehq/apps/smsbillables/management/commands/bootstrap_grapevine_gateway.py (hunk 3 above)
Comment (nickpell) [inline corehq/apps/smsbillables/management/commands/bootstrap_grapevine_gateway.py]: This is my source for the rates: http://manage.dimagi.com/default.asp?89741#513126

I'd suggest handling this in another PR (so we can tack on a new datamigration)

Comment (gcapalbo) [inline corehq/apps/smsbillables/management/commands/bootstrap_grapevine_gateway.py]: ok, tracking with http://manage.dimagi.com/default.asp?158674

**Comment by gcapalbo** [comment]
@nickpell ok, I think the only thing left here is the comment / issue about how to match the prefix.

File: corehq/apps/smsbillables/models.py (hunk 7 above)
Comment (nickpell) [inline corehq/apps/smsbillables/models.py]: fixed

File: corehq/apps/sms/phonenumbers_helper.py [hunk 9]
```diff
@@ -22,6 +22,11 @@ def parse_phone_number(number, region=None, failhard=True):
             return None
 
 
-def get_country_code(number, failhard=False):
+def get_country_code_and_national_number(number, failhard=False):
     parsed = parse_phone_number(number, failhard=failhard)
-    return parsed.country_code if parsed else None
+    if parsed:
+        country_code = parsed.country_code
+        before_national_number = "+%d" % parsed.country_code
```
Comment (gcapalbo) [inline corehq/apps/sms/phonenumbers_helper.py]: Are we sure that by the time 'number' gets here it still has the '+'? I'd be more comfortable if we didn't rely on the '+' being there, and if we instead set:

    national_number = strip_plus(number)[len(country_code):]

https://github.com/dimagi/commcare-hq/blob/sms-prefix-billing/corehq/apps/sms/util.py#L20

File: corehq/apps/smsbillables/models.py [hunk 10]
```diff
@@ -27,23 +28,43 @@ class SmsGatewayFeeCriteria(models.Model):
...
+                    ", ".join(
+                        "%(country_code)d, '%(prefix)s'" % {
+                            "country_code": criteria.country_code,
+                            "prefix": criteria.prefix,
+                        } for criteria in criteria_list
+                    )
+                )
+            criteria_list.sort(key=(lambda criteria: len(criteria.prefix)), reverse=True)
+            for criteria in criteria_list:
+                if national_number.startswith(criteria.prefix):
```
Comment (gcapalbo) [inline corehq/apps/smsbillables/models.py]: Now that national_number can be empty string, should we also check here for:

    if criteria.prefix and ...

because otherwise empty string can match empty string

Comment (gcapalbo) [inline corehq/apps/smsbillables/models.py]: And we should also check in this line for:

    if national_number and ....

because national_number can be None

File: corehq/apps/smsbillables/models.py [hunk 11]
```diff
@@ -27,23 +28,44 @@ class SmsGatewayFeeCriteria(models.Model):
...
-        - country_code (optional)
+        - country_code and prefix (optional)
         """
         all_possible_criteria = cls.objects.filter(backend_api_id=backend_api_id, direction=direction)
 
         if all_possible_criteria.count() == 0:
             return None
 
+        national_number = national_number or ""
+        def get_criteria_with_longest_matching_prefix(criteria_list):
```
Comment (dimagimon) [inline corehq/apps/smsbillables/models.py]: E301 expected 1 blank line, found 0

File: corehq/apps/sms/phonenumbers_helper.py (hunk 9 above)
Comment (nickpell) [inline corehq/apps/sms/phonenumbers_helper.py]: fixed

File: corehq/apps/smsbillables/models.py (hunk 10 above)
Comment (nickpell) [inline corehq/apps/smsbillables/models.py]: addressed offline

**Comment by gcapalbo** [comment]
@nickpell I'm ready to merge this, just waiting on the build to be safe

**Comment by nickpell** [comment]
:+1:

**Comment by nickpell** [comment]
:tada:
